presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeIdenticalWindows.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.optimizations;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.metadata.Signature;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.SymbolAllocator;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.SimplePlanRewriter;
+import com.facebook.presto.sql.planner.plan.WindowNode;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkState;
+
+/**
+ * Merge together the functions in WindowNodes that have identical WindowNode.Specifications.
+ * For example:
+ *
+ * OutputNode
+ * `--...
+ *    `--WindowNode(Specification: A, Functions: [sum(something)])
+ *       `--WindowNode(Specification: B, Functions: [sum(something)])
+ *          `--WindowNode(Specification: A, Functions: [avg(something)])
+ *             `--...
+ *
+ * Will be transformed into
+ *
+ * OutputNode
+ * `--...
+ *    `--WindowNode(Specification: A, Functions: [sum(something), avg(something)])
+ *       `--WindowNode(Specification: B, Functions: [sum(something)])
+ *          `--...
+ *
+ * This will NOT merge the functions in WindowNodes that have identical WindowNode.Specifications,
+ * but have a node between them that is not a WindowNode.
+ * In the following example, the functions in the WindowNodes with specification `A' will not be
+ * merged into a single WindowNode.
+ *
+ * OutputNode
+ * `--...
+ *    `--WindowNode(Specification: A, Functions: [sum(something)])
+ *       `--WindowNode(Specification: B, Functions: [sum(something)])
+ *          `-- ProjectNode(...)
+ *             `--WindowNode(Specification: A, Functions: [avg(something)])
```

Comment:
This is a first-step approach to solve the problem where there's a query with a lot of window functions in the select part of the query, which is a problem we have somebody actually asking us to solve.

What this approach doesn't solve for that case is when we have window functions that take constant parameters in that big list, because those introduce ProjectNodes into the chains of WindowNodes.

I did some experimentation merging the window functions into the WindowNodes both above and below the ProjectNodes, but that introduces the requirement that we change the projection list in the ProjectNode. That seems heavy-handed for this initial work.

I'm planning on the next step being to look at whether the correct approach to solving the problem of window functions that take constant parameters is to try to eliminate the ProjectNode in the plan, or whether it's to do as suggested and solve the general problem. 

That'll require a closer look at whether there are cases where we can't merge across non-WindowNodes for algebraic reasons and handling those cases appropriately.